feat(core): add ng generate schematic to convert to standalone bootstrapping APIs

crisbeto opened the pull request:
Adds the final step to the standalone schematic that converts the root component to standalone and the `bootstrapModule` to `bootstrapApplication`. Also adds a readme.

crisbeto committed:
feat(core): add ng generate schematic to convert to standalone bootstrapping APIs

Adds a new mode to the `@angular/core:standalone` schematic that allows users to switch from `bootstrapModule` to `bootstrapApplication`, as well as `provideRouter`, `provideAnimations` and `provideNoopAnimations`.

devversion commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+function addNodesToCopy(
+    targetFile: ts.SourceFile, rootNode: ts.Node, nodeLookup: NodeLookup, tracker: ChangeTracker,
+    nodesToCopy: Set<ts.Node>, languageService: ts.LanguageService): void {
+  const refs = findAllSameFileReferences(rootNode, nodeLookup, languageService);
```
We would only want to copy the declaration in practice though right? e.g.

```
const Bla = {}

providers: [Bla],
```

crisbeto replied:
Yeah in practice it'll only be declarations, but I called it a "reference" since it's a node referring to something inside the metadata.

devversion reviewed (commented):
A couple of comments. Will finish small remaining parts in a few minutes

devversion commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+ * @param modules Modules that are being imported into the new call.
+ * @param tracker Object keeping track of the changes to the different files.
+ */
+function replaceBootstrapCall(
```
Nit: the naming is quite similar to `migrateBootstrapCall` above. Maybe this can be improved.

crisbeto replied:
Yeah I was having a hard time naming it. I tweaked it slightly to `migrateBootstrapCallExpression`.

devversion commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+
+  tracker.replaceNode(
+      analysis.call, ts.factory.createCallExpression(bootstrapExpression, [], args),
+      // Note: it's important to pass in the metadata source file! Otherwise TS won't print out
```
Optional nit which I believe is easier to understand:

```suggestion
      // Note: it's important to pass in the source file where the provider expressions are taken from. [...]
```

devversion commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+            targetFile, closestExportable.name.text,
+            getRelativeImportPath(targetFile.fileName, ref.getSourceFile().fileName));
+      } else {
+        nodesToCopy.add(closestExportable);
```
I don't know exactly how importing the containing declaration would help. e.g. if a reference is part of a function. How would importing, or copying the function help? Is this actually something that would happen in practice? would be helpful to have an example.

is this for something like?

```ts
// class should be copied over then. might be a module with providers
class Bla {
  mod = ...
}

providers = [
  Bla.mod
]
```

crisbeto replied:
So this is meant to help with something like this:

```
interface SomeInterface {}

function factory(something: SomeInterface) {}

providers: [{provide: token,  useFactory: factory}]
```

In this case the schematic will identify both `factory` and `SomeInterface` as something that needs to be copied.

devversion replied:
I see. and in that case the parameter would first only point to `SomeInterface`, and you would walk up to copy the full interface. In this case the direct parent would likely work as well, but this is safer (but threw me off a little) 👍

devversion commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+ * @param languageService
+ * @param typeChecker
+ */
+function migrateImportsInBootstrapCall(
```
```suggestion
function migrateImportsForBootstrapCall(
```

devversion reviewed (approved):
LGTM overall, a couple of final comments along with the previous ones.

crisbeto committed:
docs(core): add readme for standalone migration

Updates the readme for the standalone migration to describe the process and the different transformations that are happening.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+    // If the reference is a variable, we can attempt to import it or copy it over.
+    if (variableDeclaration && variableStatement && ts.isIdentifier(variableDeclaration.name)) {
+      if (isExported(variableStatement)) {
+        tracker.addImport(
```
Yeah exactly, there's a small chance of a collision, but even if it happens, it'll be easy to fix manually.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+            targetFile, variableDeclaration.name.text,
+            getRelativeImportPath(targetFile.fileName, ref.getSourceFile().fileName));
+      } else {
+        nodesToCopy.add(variableStatement);
```
We don't need to leave a TODO, the entire variable list will be copied over to the new file and the module will be eligible for automatic deletion since it doesn't have metadata anymore.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+  };
+
+  const result = ts.transform(rootNode, [transformer]).transformed[0];
+  return hasChanged ? result : rootNode;
```
I wasn't sure how the transformer would handle it internally so I decided to play it safe.

devversion replied:
sounds reasonable

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+ * Whether a node is exported.
+ * @param node Node to be checked.
+ */
+function isExported(node: ts.Node): node is ts.Node {
```
I think that's a bit of an edge case that should be fairly easy to fix manually. I was trying to capture the most common cases.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/standalone-bootstrap.ts:
```diff
@@ -0,0 +1,630 @@
+ */
+function isExportableDeclaration(node: ts.Node): node is ts.EnumDeclaration|ts.ClassDeclaration|
+    ts.FunctionDeclaration|ts.InterfaceDeclaration|ts.TypeAliasDeclaration {
+  return ts.isEnumDeclaration(node) || ts.isClassDeclaration(node) ||
```
Variable declarations aren't exportable though, they're part of a `VariableDeclarationList` which can be.

devversion replied:
Yeah, I just called them declarations. Regardless though I wonder why `VariableDeclarationList` is not part of this list here if they are exportable? I know you treated them specially, but maybe that's not strictly necessary?

devversion replied:
yeah I've referred to the variable declaration list. Those are not part here, even though "exportable". I realize that is because you treat them specially. Likely the special code could be removed (even if the name is in `VariableDeclaration`), but this is just a nit

crisbeto reviewed (commented):
Feedback has been addressed.

jessicajaniuk commented:
This PR was merged into the repository by commit ed8f70125e722ecb8e935f8f34e89c844d9ab8ee.

cexbrayat reviewed (commented):
@crisbeto Reading the readme, I had two questions, but the PR landed before I got to submit them. If my comments make sense, I can open a PR to fix this (but maybe I misunderstood something)

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/README.md:
```diff
@@ -1,2 +1,331 @@
+// declarer.module.ts
+
+@NgModule({
+  declarations: [FooComp, BarPipe],
```
Yeah, this should be `imports`. I'll fix it up together with the other bugs.

crisbeto commented on packages/core/schematics/ng-generate/standalone-migration/README.md:
````diff
@@ -1,2 +1,331 @@
+```typescript
+// index.ts
+export {DeclarerModule} from './modules/index';
+```
````
Actually I think that only the export of `ConfigurerModule` would be left here.
